pkg/service/restore/worker.go

```diff
@@ -45,37 +45,160 @@ type worker struct {
+	// in the cluster (e.g. tombstone_gc mode, views definition, etc).
+	if err := w.initUnits(ctx, w.target.locationInfo); err != nil {
 		return errors.Wrap(err, "init units")
 	}
 	return errors.Wrap(w.initViews(ctx), "init views")
 }
 
-func (w *worker) initTarget(ctx context.Context, properties json.RawMessage) error {
-	t := defaultTarget()
-	if err := json.Unmarshal(properties, &t); err != nil {
-		return err
+func (w *worker) getLocationInfo(ctx context.Context, target Target) ([]LocationInfo, error) {
+	var result []LocationInfo
+
+	nodeStatus, err := w.client.Status(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "get status")
+	}
+
+	sourceDC2TargetDCMap, targetDC2SourceDCMap := target.DCMappings, reverseMap(target.DCMappings)
+
+	for _, l := range target.Location {
+		nodes, err := w.getNodesWithAccess(ctx, nodeStatus, l, len(target.DCMappings) > 0)
+		if err != nil {
+			return nil, errors.Wrap(err, "getNodesWithAccess")
+		}
+
+		manifests, err := w.getManifestInfo(ctx, nodes[0].Addr, l, target.SnapshotTag)
+		if err != nil {
+			return nil, errors.Wrap(err, "getManifestInfo")
+		}
+		if len(manifests) == 0 {
+			return nil, errors.Errorf("no snapshot with tag %s", target.SnapshotTag)
+		}
+
+		manifests = filterManifests(manifests, sourceDC2TargetDCMap)
+		locationDCs := collectDCsFromManifests(manifests)
+		dcHosts := hostsByDC(nodes, targetDC2SourceDCMap, locationDCs)
+
+		result = append(result, LocationInfo{
+			DC:       locationDCs,
+			DCHosts:  dcHosts,
+			Manifest: manifests,
+			Location: l,
+		})
+	}
+
+	return result, nil
+}
+
+// From map[k]v to map[v]k.
+func reverseMap(m map[string]string) map[string]string {
+	result := make(map[string]string, len(m))
+	for k, v := range m {
+		result[v] = k
+	}
+	return result
+}
+
+func (w *worker) getNodesWithAccess(
+	ctx context.Context,
+	nodeStatus scyllaclient.NodeStatusInfoSlice,
+	loc backupspec.Location,
+	useLocationDC bool,
+) (scyllaclient.NodeStatusInfoSlice, error) {
+	if useLocationDC && loc.Datacenter() != "" {
+		nodeStatus = nodeStatus.Datacenter([]string{loc.Datacenter()})
+	}
+
+	nodes, err := w.client.GetNodesWithLocationAccess(ctx, nodeStatus, loc.RemotePath(""))
+	if err != nil {
+		if strings.Contains(err.Error(), "NoSuchBucket") {
+			return nil, errors.Errorf("specified bucket does not exist: %s", loc)
+		}
+		return nil, errors.Wrapf(err, "location %s is not accessible", loc)
+	}
+	if len(nodes) == 0 {
+		return nil, fmt.Errorf("no nodes with location %s access", loc)
+	}
+	return nodes, nil
+}
+
+// hostsByDC creates map of which hosts are responsible for which DC, also applies DCMappings if available.
+func hostsByDC(nodes scyllaclient.NodeStatusInfoSlice, targetDC2SourceDCMap map[string]string, locationDCs []string) map[string][]string {
+	dc2HostsMap := map[string][]string{}
+	// When --dc-mapping is not set all nodes (or nodes from location.DC) with access to the location can handle all DCs (or location.DC) from it
```

Comment:
Is the second `(or location.DC)` needed?
Setting location's DC means that nodes from this DC will be restoring this location.
It doesn't mean that the only data center restored from this location will be this DC.

Also a small nit, this comment could be broken into multiple lines.